### PR TITLE
Save button only enabled if form is filled out correctly

### PR DESCRIPTION
**Pull request by melanieretter**
The save button in create/update dialogs is only enabled if all mandatory fields are filled out and if changes exists (in update case)

savebuttondisabled

### CODE REVIEW

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/artifacts/smtable/SoftwareModuleAddUpdateWindow.java [hunk 1]
```diff
@@ -209,17 +228,32 @@ private void createWindow() {
...
+                SPUIDefinitions.CREATE_UPDATE_WINDOW, this, event -> saveOrUpdate(), event -> closeThisWindow(), null,
+                getMandatoryFields(formLayout), geEditedFields(), i18n);
         window.getButtonsLayout().removeStyleName("actionButtonsMargin");
-        nameTextField.focus();
+        typeComboBox.focus();
     }
 
-    private void addDescriptionTextChangeListener() {
-        descTextArea.addTextChangeListener(event -> window.setSaveButtonEnabled(hasDescriptionChanged(event)));
+    private Map<String, Boolean> getMandatoryFields(final FormLayout formLayout) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/artifacts/smtable/SoftwareModuleAddUpdateWindow.java]: The method is duplicated

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/CommonDialogWindow.java [hunk 2]
```diff
@@ -72,28 +93,207 @@
...
+     * @param originalValue
+     *            original Value of the current field
+     */
+    public void checkMandatoryEditedValue(final ValueChangeEvent event, final AbstractComponent component,
+            final String originalValue) {
+        final boolean isChangedValueNotNull = event.getProperty().getValue() != null;
+        if (requiredFields.containsKey(component.getId())) {
+            setRequiredFieldChangeValue(component, isChangedValueNotNull);
+        }
+        if (event.getProperty().getValue() != null) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/CommonDialogWindow.java]: sonar rule: https://waterfall.auckland.ac.nz/sonar/rules/show/groovy:org.codenarc.rule.convention.ConfusingTernaryRule?layout=false

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/CommonDialogWindow.java [hunk 3]
```diff
@@ -167,4 +392,12 @@ public HorizontalLayout getButtonsLayout() {
         return buttonsLayout;
     }
 
+    public Map<String, Boolean> getRequiredFields() {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/common/CommonDialogWindow.java]: Is not used or?

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/distributions/disttype/CreateUpdateDistSetTypeLayout.java [hunk 4]
```diff
@@ -237,44 +245,87 @@ private void buildSourceTable() {
...
+        originalSelectedTableContainer.addContainerProperty(DIST_TYPE_DESCRIPTION, String.class, "");
+        originalSelectedTableContainer.addContainerProperty(DIST_TYPE_MANDATORY, CheckBox.class, null);
     }
 
     @SuppressWarnings("unchecked")
     private void addSMType() {
 
         final Set<Long> selectedIds = (Set<Long>) sourceTable.getValue();
-        if (null != selectedIds && !selectedIds.isEmpty()) {
+        if (selectedIds != null && !selectedIds.isEmpty()) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/distributions/disttype/CreateUpdateDistSetTypeLayout.java]: Would prefer early return

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/distributions/disttype/CreateUpdateDistSetTypeLayout.java [hunk 5]
```diff
@@ -237,44 +245,87 @@ private void buildSourceTable() {
...
+            window.updateRequiredFields(selectedTable.getId(), hasContentChanged());
         }
     }
 
     private void removeSMType() {
 
         @SuppressWarnings("unchecked")
         final Set<Long> selectedIds = (Set<Long>) selectedTable.getValue();
-        if (null != selectedIds && !selectedIds.isEmpty()) {
+        if (selectedIds != null && !selectedIds.isEmpty()) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/distributions/disttype/CreateUpdateDistSetTypeLayout.java]: Would prefer early return

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/distributions/disttype/CreateUpdateDistSetTypeLayout.java [hunk 6]
```diff
@@ -237,44 +245,87 @@ private void buildSourceTable() {
...
+            if (!originalSelectedTableContainer.containsId(itemId)) {
+                window.updateEditedFields(selectedTable.getId(), Boolean.TRUE);
+                return Boolean.TRUE;
+            }
+        }
+
+        for (final Iterator itemIterator = originalSelectedTableContainer.getItemIds().iterator(); itemIterator
+                .hasNext();) {
+            final long itemId = (Long) itemIterator.next();
+            if (selectedTableContainer.size() > 0 && !selectedTableContainer.containsId(itemId)) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/distributions/disttype/CreateUpdateDistSetTypeLayout.java]: Is the size check necessary?

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/distributions/disttype/CreateUpdateDistSetTypeLayout.java [hunk 7]
```diff
@@ -584,37 +622,66 @@ private DistributionSetType fetchDistributionSetType(final String distTypeName)
     private void addTargetTableforUpdate(final SoftwareModuleType swModuleType, final boolean mandatory) {
 
         Item saveTblitem;
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/distributions/disttype/CreateUpdateDistSetTypeLayout.java]: move it it in the if

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/distributions/disttype/CreateUpdateDistSetTypeLayout.java [hunk 8]
```diff
@@ -584,37 +622,66 @@ private DistributionSetType fetchDistributionSetType(final String distTypeName)
...
-        if (null != selectedTablecontainer) {
-            saveTblitem = selectedTablecontainer.addItem(swModuleType.getId());
+        if (selectedTableContainer != null) {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/distributions/disttype/CreateUpdateDistSetTypeLayout.java]: Prefer early return

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/distributions/disttype/CreateUpdateDistSetTypeLayout.java [hunk 9]
```diff
@@ -584,37 +622,66 @@ private DistributionSetType fetchDistributionSetType(final String distTypeName)
...
     @Override
     public void createWindow() {
         reset();
         window = SPUIComponentProvider.getWindow(i18n.get("caption.add.type"), null,
-                SPUIDefinitions.CREATE_UPDATE_WINDOW, this, this::save, this::discard, null);
+                SPUIDefinitions.CREATE_UPDATE_WINDOW, this, this::save, this::discard, null, getMandatoryFields(),
+                getEditedFields(), i18n);
+    }
+
+    private Map<String, Boolean> getMandatoryFields() {
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/distributions/disttype/CreateUpdateDistSetTypeLayout.java]: Looks duplicated

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/layouts/AbstractCreateUpdateTagLayout.java [hunk 10]
```diff
@@ -105,10 +109,35 @@
...
         reset();
-        setWindow(SPUIComponentProvider.getWindow(i18n.get("caption.add.tag"), null,
-                SPUIDefinitions.CREATE_UPDATE_WINDOW, this, this::save, this::discard, null));
+        setWindow(
+                SPUIComponentProvider.getWindow(i18n.get("caption.add.tag"), null, SPUIDefinitions.CREATE_UPDATE_WINDOW,
+                        this, this::save, this::discard, null, getMandatoryFields(), getEditedFields(), i18n));
+    }
+
+    private Map<String, Boolean> getMandatoryFields() {
+        final Map<String, Boolean> requiredFields = new HashMap<>();
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/layouts/AbstractCreateUpdateTagLayout.java]: Duplicated :)

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/layouts/AbstractCreateUpdateTagLayout.java [hunk 11]
```diff
@@ -320,6 +344,7 @@ protected void optionValueChanged(final ValueChangeEvent event) {
             comboLayout.removeComponent(comboLabel);
             comboLayout.removeComponent(tagNameComboBox);
         }
+        window.reset();
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/layouts/AbstractCreateUpdateTagLayout.java]: 2 times reset? And no null check?

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/layouts/CreateUpdateTypeLayout.java [hunk 12]
```diff
@@ -45,14 +41,15 @@
     protected String createTypeStr;
     protected String updateTypeStr;
     protected TextField typeKey;
+    protected String originalTypeKey;
```
Comment (SirWayne) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/layouts/CreateUpdateTypeLayout.java]: why do you not use a private member?

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/distributions/disttype/CreateUpdateDistSetTypeLayout.java [hunk 13]
```diff
@@ -651,4 +710,14 @@ protected void createOptionGroup(final boolean hasCreatePermission, final boolea
         optiongroup.setId(SPUIDefinitions.CREATE_OPTION_GROUP_DISTRIBUTION_SET_TYPE_ID);
     }
 
+    // @Override
```
Comment (Jonkno) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/distributions/disttype/CreateUpdateDistSetTypeLayout.java]: Remove

File: hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/utils/CommonDialogWindowHelper.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+import com.vaadin.ui.AbstractField;
+import com.vaadin.ui.Component;
+import com.vaadin.ui.FormLayout;
+
+public class CommonDialogWindowHelper {
```
Comment (Jonkno) [inline hawkbit-ui/src/main/java/org/eclipse/hawkbit/ui/utils/CommonDialogWindowHelper.java]: Add Javadoc

**Comment by melanieretter** [comment]
Replace with this pull request: https://github.com/eclipse/hawkbit/pull/233

**Pull request closed without merging**